### PR TITLE
Implementation of missing block-liquid detection

### PR DESCRIPTION
**Pull request by CreativeMD**
Recently, I came across `Block.isEntityInsideMaterial` and `Block.isAABBInsideMaterial`, which are a great opportunity for me to add swimming physics for LittleTiles ([video](https://youtu.be/cEQqUSXbp_s?t=55s)). Unfortunately I rand into two issues.

**First**
To my surprise lava did not work. Looking at the actual code inside `Entity.class` ...
```java
public boolean isInLava()
{
    return this.world.isMaterialInBB(this.getEntityBoundingBox().grow(-0.10000000149011612D, -0.4000000059604645D, -0.10000000149011612D), Material.LAVA);
}
```
... it turns out that `world.isMaterialInBB` is still using the old way of detecting a material, so I changed it to use `Block.isAABBInsideMaterial` instead, which btw is never used at all.

**Second**
Furthermore there are some wonky physics, which would make you go up as soon as you swim against a block. After some further investigation I found the source of the problem ...
```java
private boolean isLiquidPresentInAABB(AxisAlignedBB bb)
{
    return this.world.getCollisionBoxes(this, bb).isEmpty() && !this.world.containsAnyLiquid(bb);
}
```
`world.containsAnyLiquid` will use the state to check whether a block is a liquid or not. So I created a new method inside `Block.class` which should allow a block to check whether a given bounding box is inside a liquid or not: `isAABBInsideLiquid`. I implemented it in the same way as the others to keep consistency.

### CODE REVIEW

**Comment by mezz** [comment]
`isAABBInsideLiquid` seems like it's equivalent to calling `isAABBInsideMaterial` with `Material#WATER` or `Material#LAVA`. Can you explain a bit more why it's needed? I think the rest makes sense.

**Comment by CreativeMD** [comment]
There might be more liquids than `Material#WATER` and `Material#LAVA` by other mods. So you would have to go through all possible liquids and call `isAABBInsideMaterial` for all of them. I don't think this is good for performance, so I added this extra method.

File: patches/minecraft/net/minecraft/world/World.java.patch [hunk 1]
```diff
@@ -541,7 +541,24 @@
...
-@@ -2011,6 +2163,11 @@
+@@ -1972,6 +2124,16 @@
+                 {
+                     IBlockState iblockstate1 = this.func_180495_p(blockpos$pooledmutableblockpos.func_181079_c(l3, i4, j4));
+ 
++                    Boolean result = iblockstate1.func_177230_c().isAABBInsideLiquid(this, blockpos$pooledmutableblockpos, p_72953_1_);
++
++                    if (result != null)
++                    {
++                        if (result == true)
```
Comment (liach) [inline patches/minecraft/net/minecraft/world/World.java.patch]: `if (result)`

Comment (CreativeMD) [reply patches/minecraft/net/minecraft/world/World.java.patch]: I know, but i wanted to keep it the same as that one https://github.com/MinecraftForge/MinecraftForge/blob/1.12.x/patches/minecraft/net/minecraft/world/World.java.patch#L561

Comment (liach) [reply patches/minecraft/net/minecraft/world/World.java.patch]: You mean #2429? I think that is just a bad practice (`result != null && result.booleanValue()` is even clearer). Back then, people rarely comment on diffs, too, and I guess it's why that patch was included.

Comment (liach) [reply patches/minecraft/net/minecraft/world/World.java.patch]: By the way, you should release the pooled mutable block position before returning (or jumping to the next loop if the mutable block position is retained within the loop)

**Review by liach — APPROVED** [review]

**Comment by mezz** [comment]
There is `Material#isLiquid` to check if the material is liquid, so I think the old method is still good.

**Comment by CreativeMD** [comment]
This doesn't make sense in my case as my default `blockMaterial` is `Material#ROCK`, but there might be liquids placed inside. It's basically the whole point of those hooks, to fake a material. I don't see a different way to solve this, except of the messy solution to check for all the liquids.

**Comment by liach** [comment]
Are you checking for your water tanks?

**Comment by CreativeMD** [comment]
No, I need it for LittleTiles, which allows you to build smaller pieces of a block. Including water and lava ([water example](https://youtu.be/cEQqUSXbp_s?t=55s)). Maybe it will support more fluids in the future. As I said before those hooks should allow a block to simulate all the liquids (original pr was created by AlgorithmX2 who is the author of C&B).

**Comment by bs2609** [comment]
A better alternative to testing for materials individually would be to instead have a method that returns a `Set<Material>` which can then be tested for particular materials, or fluids, or anything else.

**Comment by CreativeMD** [comment]
I could, but i wanted to keep things simple.

**Comment by mezz** [comment]
@AlgorithmX2 can you look this over? I'd like to get your thoughts on this one.

**Comment by AlgorithmX2** [comment]
I think @CreativeMD is right about this, it looks like the call to isAABBInsideMaterial was lost at some point.

Original PR: https://github.com/MinecraftForge/MinecraftForge/pull/2429/commits/3c4a2b0f4289646fecbc9785add00272bf3ef930

**Comment by mezz** [comment]
Please update this PR to resolve merge conflicts.

**Comment by CreativeMD** [comment]
I love git ... f***. I will create a new pr, sorry :(

**Comment by diesieben07** [comment]
For the future:
This is why you make a separate branch when working on a PR. If something goes wrong, you can nuke the branch, create a new one in it's place and then force-push. Then you do not need a new PR.

**Comment by CreativeMD** [comment]
Yes, this might be a good idea. I will do it different the next time...

**Pull request closed without merging**